core/src/main/java/org/bitcoinj/base/BitcoinNetwork.java

```diff
@@ -76,17 +77,49 @@ public Coin maxMoney() {
         return MAX_MONEY;
     }
 
+    /**
+     * Get the {@code BitcoinNetwork} from a <i>validated</i> name String
+     * @param nameString A name string (e.g. "MAIN", "TEST", "SIGNET")
+     * @return The matching enum
+     * @throws IllegalArgumentException if there is no matching enum
+     */
+    public static BitcoinNetwork of(String nameString) {
+        return find(nameString)
+                .orElseThrow(() -> new IllegalArgumentException("Unrecognized network name : " + nameString));
+    }
+
+    /**
+     * Find the {@code BitcoinNetwork} from a name String
+     * @param nameString A name string (e.g. "MAIN", "TEST", "SIGNET")
+     * @return An {@code Optional} containing the matching enum or empty
+     */
+    public static Optional<BitcoinNetwork> find(String nameString) {
+        return Arrays.stream(values())
+                .filter(n -> n.toString().equals(nameString.toUpperCase()))
+                .findFirst();
+    }
+
     /**
      * Get the correct enum for a network id string
      * <p>
      * Note: UNITTEST is not supported as an enum
      * @param idString specifies the network
      * @return the enum
+     * @throws IllegalArgumentException if there is no matching enum
+     */
+    public static BitcoinNetwork fromId(String idString) {
+        return findById(idString)
+                .orElseThrow(() -> new IllegalArgumentException("Illegal network ID: " + idString));
+    }
+
+    /**
+     * Find the {@code BitcoinNetwork} from an ID String
+     * @param idString specifies the network
+     * @return An {@code Optional} containing the matching enum or empty
      */
-    public static BitcoinNetwork of(String idString) {
```

Comment:
I'm not sure if it's a good idea to change this "by ID" constructor from strict `of` to lenient `find`. After all, this is an ID which is likely not user facing (other than the freshly introduced "name"). Maybe restore the old error handling and name it `ofId()`, or keep the `of()` name and use `ofName()` for the other constructor?

Reply:
The diff is a little hard to read, the strict 'of' became a strict `fromId` but the diff shows that as inserted above it.

This PR results in 4 static factory methods:

of(name)    -- strict: either returns a `BitcoinNetwork` or throws exception
fromId(id)   -- strict: either returns a `BitcoinNetwork` or throws exception
find(name)  -- lenient: returns an Optional<BitcoinNetwork>
findById(id) -- lenient: returns an Optional<BitcoinNetwork> 

I think this covers all the use cases. I have used network ID strings in configuration files for server apps and think others may do similar things. In that case you might view the ID string as "user facing". I thought the `of` should use the "name" and be the inverse of `.toString()` or `.name()`. We could rename `fromId()` to `ofId()` I suppose.